File: lua/bank-account/bank-account.lua
local BankAccount = {
  new = function(self)
    local o = {
      _open = true,
      _balance = 0
    }

    setmetatable(o, self)
    self.__index = self

    return o
  end,

  balance = function(self)
    return self._balance
  end,

  deposit = function(self, amount)
    if amount <= 0 then
      error("amount must be positive")
    end
    if not self._open then
      error("operation on closed account")
    end
    self._balance = self._balance + amount
  end,

  withdraw = function(self, amount)
    if amount <= 0 or amount >= self._balance then
      error("amount must be positive and less than balance")
    end
    if not self._open then
      error("operation on closed account")
    end

    self._balance = self._balance - amount
  end,

  close = function(self)
    self._open = false
  end
}



return BankAccount

File: python/bank-account/bank_account.py
class BankAccount:
    _open = False
    _balance = 0

    def _assert_open(self):
        if not self._open:
            raise ValueError("account not open")

    def get_balance(self):
        self._assert_open()
        return self._balance

    def open(self):
        if self._open:
            raise ValueError("account already open")
        self._open = True

    def deposit(self, amount):
        self._assert_open()

        if amount < 0:
            raise ValueError("amount must be greater than 0")

        self._balance += amount

    def withdraw(self, amount):
        self._assert_open()

        if amount > self._balance:
            raise ValueError("amount must be less than balance")

        if amount < 0:
            raise ValueError("amount must be greater than 0")

        self._balance -= amount

    def close(self):
        self._assert_open()
        self._open = False
        self._balance = 0
